packages/animations/lib/src/utils/composite_animation_widget.dart

```diff
@@ -0,0 +1,253 @@
+  ///
+  /// This widget will transition in and out as driven by [animation].
+  final Widget child;
+
+  /// A function that wraps [child] with an animation set define how the
+  /// child appears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes forward.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.completed] (1.0).
+  ///
+  /// Must not change subtree levels.
+  final _EnterTransitionBuilder enterTransitionBuilder;
+
+  /// A function that wraps [child] with an animation set define how the
+  /// child disappears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes reverse.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.dismissed] (0.0).
+  ///
+  /// Must not change subtree levels.
+  final _ExitTransitionBuilder exitTransitionBuilder;
+
+  /// Indicates whether the animation values will be flipped.
+  ///
+  /// Flipping animation values will also flips visible states in defined
+  /// transition builders.
+  ///
+  /// So when flipped, The visible state of widgets in [enterTransitionBuilder]
+  /// and [exitTransitionBuilder] will be at [AnimationStatus.dismissed] (0.0)
+  /// and [AnimationStatus.completed] (1.0) respectively.
+  final bool flip;
+
+  /// Sync visible state of [exitTransitionBuilder] to [enterTransitionBuilder].
+  /// So when true, visible state of [exitTransitionBuilder] must be the same as
+  /// [enterTransitionBuilder].
+  ///
+  /// Set it to true when the [exitTransitionBuilder]'s animation widget
+  /// has visible state same as the animation widgets defined in
+  /// [enterTransitionBuilder].
+  ///
+  /// Defined true at [FadeScaleTransition] since it only used one animation.
+  final bool syncVisibleStates;
+
+  @override
+  _CompositeAnimationWidgetState createState() =>
+      _CompositeAnimationWidgetState();
+}
+
+class _CompositeAnimationWidgetState extends State<CompositeAnimationWidget> {
+  AnimationStatus _effectiveAnimationStatus;
+  Animation<double> _enterAnimation;
+  Animation<double> _exitAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.animation.addStatusListener(_animationListener);
+    _effectiveAnimationStatus = widget.animation.status;
+    _setStartingAnimationValues(widget.flip);
+  }
+
+  void _animationListener(AnimationStatus animationStatus) {
+    _effectiveAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveAnimationStatus,
+      current: animationStatus,
+    );
+    // change the animation values based on effectiveAnimationStatus.
+    _setStartingAnimationValues(widget.flip);
+  }
+
+  // This is called at [initState], [didUpdateWidget], and [_animationListener]
+  void _setStartingAnimationValues(bool flip) {
+    final Animation<double> baseAnimation =
```

Comment:
this method is hard to follow because _enterAnimation and _exitAnimation change their value fairly often. Can you refactor this to make the flow clearer?

E.g. maybe start with the switch and then in each case configure _enterAnimation/_exitAnimation based on the value of flip and syncVisibleStates?

Reply:
Ok, And I think ill make an additional `_setStartingAnimationValues` function parameters for `effectiveAnimationStatus` and `syncVisibleState` or access them directly to widget property?